### PR TITLE
PCHR-2719: Leave Widget Balances

### PR DESCRIPTION
**Pull request by reneolivo**
## Overview
This PR displays the future balance for leave requests where the user has entitlements for.

## Before
download 83

## After
anim

### Contact ID

The contact ID comes from Contact Summary Settings module, which provides a `settings.contactId` value of the contact whose details are being examined. 

This contact id is passed to the leave widget:
```html
<leave-widget contact-id="ContactSummaryCtrl.contactId"></leave-widget>
```

### Loading

The `crm-loading` directive from reqangular is used to hide the component's content while dependencies are loading.

The `crm-loading` is ON when `.loading.childComponents` and `.loading.component` are true
```html
<crm-loading show="!leaveWidget.loading.component && !leaveWidget.loading.childComponents">
  ...
</crm-loading>
```

`.loading.component` is set to TRUE while the component loads Absence Types and the current Absence Period, both of which are dependencies needed by all child components.

`.loading.childComponents` is set to TRUE when child components fire the `LeaveWidget::childIsLoading` event, and is set to FALSE when the last `LeaveWidget::childIsReady` is fired.
([good starting point to see this in the code](https://github.com/civicrm/civihr/pull/2206/files#diff-d127156ec6d73b695fee22489c6a8618R46))

### Leave Widget Balance

The `leave-widget-balance` component is a child of the `leave-widget`. It requires the Absence Types, the Contact Id, and the Current Absence Period to be passed to it:

```html
<leave-widget-balance
        absence-types="leaveWidget.absenceTypes"
        contact-id="leaveWidget.contactId"
        current-absence-period="leaveWidget.currentAbsencePeriod"
></leave-widget-balance>
```

This child component emits the `LeaveWidget::childIsReady` when the expected bindings are ready to use, since some of them come from promises they may not be ready when the component first starts. It also needs to wait for Entitlements to load.

Entitlements are loaded using the following parameters:
```js
return Entitlement.all({
  contact_id: vm.contactId,
  period_id: vm.currentAbsencePeriod.id
}, true)
```

The contact id and period id come from the component bindings, the **true** parameter means that the entitlement remainders are also loaded in this request. This not only brings the data for the entitlement, but also the current and future balance, the future being the one we are interested in since it's defined by approved and waiting for approval requests.

Since the entitlements don't come with the absence type information attached to them, the `mapAbsenceTypesWithTheirEntitlements` is used to add the remainder entitlement for each absence type.

```js
var absenceType = _.find(vm.absenceTypes, function (type) {
  return +type.id === +entitlement.type_id;
});
```
finds the absence type for this particular entitlement.

```js
absenceType = _.assign({
  balance: entitlement.remainder.future
}, absenceType);
```
creates a copy of the absence type, with the balance property being the entitlement's future remainder.

### Tests

The following tests were added:

* Contact Summary Ctrl's Contact Id - it tests that the contact id is available for the Contact Summary view.
* Leave Widget loading dependencies, absence types, and current absence period.
* Leave Widget loading child components.
* Leave Widget Balance waiting for bindings.
* Leave Widget Balance loading entitlements.

Both Contact Summary and L&A extensions tests ran successfully.

### CODE REVIEW

**Comment by igorpavlov** [comment]
@Rene could you please clarify which exact types of tests were run?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-balance.component.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      return sharedSettings.sharedPathTpl + 'components/leave-widget/leave-widget-balance.html';
+    }]
+  });
+
+  leaveWidgetBalanceController.$inject = ['$scope', 'Entitlement'];
+
+  function leaveWidgetBalanceController ($scope, Entitlement) {
+    var entitlements;
+    var vm = this;
+    vm.$onChanges = $onChanges;
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-balance.component.js]: An extra line needed before `vm.$onChanges = $onChanges;` ?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-balance.component.js [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      return vm.absenceTypes && vm.currentAbsencePeriod && vm.contactId;
+    }
+
+    /**
+     * Loads all the component dependencies (entitlements in this case) and
+     * emits a child is ready event.
+     */
+    function loadDependencies () {
+      loadEntitlements()
+      .then(function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-balance.component.js]: I think in this case .then can go right after loadEntitlements()

```js
loadEntitlements().then(function () {
//...
```

up to you.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-balance.component.js [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        $scope.$emit('LeaveWidget::childIsReady');
+      });
+    }
+
+    /**
+     * Loads entitlements for the selected user and absence period and maps
+     * the entitlements to their corresponding absence type.
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-balance.component.js]: That func returns promise, please add in the func description.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-balance.component.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    function areBindingsReady () {
+      return vm.absenceTypes && vm.currentAbsencePeriod && vm.contactId;
+    }
+
+    /**
+     * Loads all the component dependencies (entitlements in this case) and
+     * emits a child is ready event.
+     */
+    function loadDependencies () {
+      loadEntitlements()
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget-balance.component.js]: I think this func should return promise (why not?).

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js [hunk 5]
```diff
@@ -1,19 +1,108 @@
 /* eslint-env amd */
 
 define([
-  'leave-absences/shared/modules/components'
-], function (components) {
+  'common/lodash',
+  'leave-absences/shared/modules/components',
+  'leave-absences/shared/models/absence-period.model',
+  'leave-absences/shared/models/absence-type.model',
+  './leave-widget-balance.component'
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js]: Shouldn't the path be specified in the conf file? Just wondering.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js [hunk 6]
```diff
@@ -1,19 +1,108 @@
...
+], function (_, components) {
   components.component('leaveWidget', {
+    bindings: {
+      contactId: '='
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js]: = means leaveWidget can change the contactId?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js [hunk 7]
```diff
@@ -1,19 +1,108 @@
...
+  function leaveWidgetController ($log, $scope, AbsencePeriod, AbsenceType) {
+    var childComponents = 0;
+    var vm = this;
+
+    vm.absenceTypes = [];
+    vm.currentAbsencePeriod = null;
+    vm.loading = { childComponents: false, component: true };
+
+    /**
+     * Initializes the component by watching for events, and loading
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js]: Do we write func comments to init() ?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js [hunk 8]
```diff
@@ -1,19 +1,108 @@
...
+    /**
+     * Watches for child components loading and ready events.
+     */
+    function initWatchers () {
+      $scope.$on('LeaveWidget::childIsLoading', childComponentIsLoading);
+      $scope.$on('LeaveWidget::childIsReady', childComponentIsReady);
+    }
+
+    /**
+     * Loads absence types and the current absence period. When
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js]: Shouldn't this func return a promise (why not?)

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-widget/leave-widget-balance.component.spec.js [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  'leave-absences/shared/components/leave-widget/leave-widget-balance.component',
+  'mocks/apis/entitlement-api-mock'
+], function (_) {
+  describe('LeaveWidgetBalance', function () {
+    var $componentController, $provide, $rootScope, $scope,
+      currentAbsencePeriod, absenceTypes, ctrl, Entitlement;
+    var contactId = 101;
+
+    beforeEach(module('leave-absences.components', 'leave-absences.mocks',
+    function (_$provide_) {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-widget/leave-widget-balance.component.spec.js]: I would indent `function (_$provide_) {` with 2 spaces, it was a bit difficult to read:
```js
beforeEach(module('leave-absences.components', 'leave-absences.mocks',
  function (_$provide_) {
    $provide = _$provide_;
  }));
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-widget/leave-widget-balance.component.spec.js [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+    describe('on init', function () {
+      it('fires a leave widget child is loading event', function () {
+        expect($scope.$emit).toHaveBeenCalledWith(
+          'LeaveWidget::childIsLoading');
+      });
+    });
+
+    describe('bindings', function () {
+      describe('when absence types, current absence period, and contact id are binded', function () {
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-widget/leave-widget-balance.component.spec.js]: binded = bound

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-widget/leave-widget-balance.component.spec.js [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+            expect($scope.$emit).toHaveBeenCalledWith(
+              'LeaveWidget::childIsReady');
+          });
+        });
+      });
+    });
+
+    function controllerOnChanges (bindingName, bindingValue) {
+      var changes = {};
+      ctrl[bindingName] = bindingValue;
```
Comment (igorpavlov) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-widget/leave-widget-balance.component.spec.js]: Needs an empty line before `ctrl[bindingName] = bindingValue;`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js (hunk 6 above)
Comment (reneolivo) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js]: it should be < !!!

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js (hunk 7 above)
Comment (reneolivo) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js]: Not sure, is it wrong?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-widget/leave-widget-balance.component.spec.js (hunk 10 above)
Comment (reneolivo) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-widget/leave-widget-balance.component.spec.js]: Engrish is hard

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js (hunk 5 above)
Comment (reneolivo) [reply uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-widget/leave-widget.component.js]: Paths can be relative.

**Comment by reneolivo** [comment]
### New update

The leave balance is displayed using the `timeUnitApplier` filter.

#### After

download 84


#### Technical details

The filter is applied in this manner: `{{ type.balance | timeUnitApplier:'days' }}`. "days" is used as a constant since this is the only supported option until Leave in Hours is completed.

A new module was created, `leave-absences.components.leave-widget` to replace `leave-absences.components`. This was done in order to include specific components for this widget that are not available to other components, such as the `timeUnitApplier` filter. This helps reduce file size in the L&A side, since it avoids adding the filter by default.

**Pull request merged**